src/ws/cockpitauth.c

```diff
@@ -1469,14 +1468,20 @@ cockpit_auth_login_async (CockpitAuth *self,
 
   application = cockpit_auth_parse_application (path, NULL);
 
+  JsonObject *metadata = connection ?
+                           g_object_get_qdata (G_OBJECT (connection),
+                                               g_quark_from_static_string ("metadata"))
+                         :
+                           NULL;
+
   /* If the client sends a TLS certificate to cockpit-tls, treat this as a
    * definitive login type, and don't just silently fall back to other types */
-  if (https_instance_has_certificate_file (NULL, 0) != -1)
+  const gchar *client_certificate;
+  if (metadata && cockpit_json_get_string (metadata, "client-certificate", NULL, &client_certificate))
     {
       g_debug ("TLS connection has peer certificate, using tls-cert auth type");
       type = g_strdup ("tls-cert");
-      /* don't send any actual authorization here; we don't want to put any trust in data sent from cockpit-ws */
-      authorization = g_strdup ("tls-cert");
+      authorization = g_strdup (client_certificate);
```

Comment:
What does that buy us eventually? c-session does not read that at the moment (as the commit message points out, "continued to be ignored"), but the original comment still stands -- cockpit-session can't trust anything from cockpit-ws. Client certs are public, and thus their fingerprints are predictable, so a hacked cockpit-ws could send any value to cockpit-session.

I'd feel much better about connection_create_metadata() just sending `client-certificate: true`, so that it can tell ws to use tls-cert mode. But that should not be passed on to c-session.

Reply:
maybe the name is misleading: this is not the cert, but rather the basename of the file in /run/cockpit/tls which contains the cert.

Reply:
Of course, but same difference -- as I mentioned, the fingerprint (aka cert path basename) is also predictable. As that information passed to c-session is untrustworthy, we shouldn't pass it so that we can be sure c-session won't ever start to use it.

Reply:
FTR: it's similar to a password, c-session will still validate it via cgroup membership and content comparison.